add spectrum class

drammock opened the pull request:
Progress so far:

- [x] `.compute_psd()` method for Raw, Epochs, Evoked yields Spectrum object
- [x] supports method='welch' and method='multitaper'
- [x] supports unaggregated welch via `average=False`
- [x] supports unaggregated multitaper via `output='complex'`
- [ ] ~~support passing a callable to aggregate welch/multitaper estimates~~ #11093
- [x] properties ch_names, freqs, method (welch/multitaper)
- [x] private property `._dims`, e.g., `('epoch', 'channel', 'freq', 'taper')` for an unaggregated multitaper derived from Epochs
- [x] save to / load from HDF5 (`spectrum.save(fname)` / `mne.time_frequency.read_spectrum(fname)`) for Spectrum object
- [x] save/load for EpochsSpectrum object
- [x] `__eq__` method for comparisons / testing
- [x] `__getitem__` works for raw/epochs/evoked (raw/evoked-derived spectra work like np.array; epochs-derived spectra work like epochs: select by condition name, pandas query, etc)
- [x] `.pick()` method
- [x] `.units()` method
- [x] `.to_data_frame()` method (works with raw/epochs/evoked, agged/unagged estimates)
- [x] `.plot()` method with equivalent API to old `inst.plot_psd()` method
- [x] `.plot_topomap()` method
- [x] `.plot_topo()` method
- [x] tests
- [x] tutorial
- [x] deprecations
- [x] changelog

```py
import os
import mne

sample_data_folder = mne.datasets.sample.data_path()
sample_data_raw_file = os.path.join(sample_data_folder, 'MEG', 'sample',
                                    'sample_audvis_raw.fif')
raw = mne.io.read_raw_fif(sample_data_raw_file, verbose=False, preload=False)

spectrum = raw.compute_psd()
spectrum.plot()
```
yields this:
![Figure_1](https://user-images.githubusercontent.com/1810515/149236966-d3240a8d-5dc4-48c2-a687-9b8bea6a7443.png)

`spectrum` has properties `_data`, `freqs`, `ch_names`, `method` (welch or multitaper), an `info` dict, and `spec._dims` (description of the data dims). For example

```py
In [8]: unagg_spec = raw.to_spectrum(average=None)
In [9]: unagg_spec._dims
Out[9]: ('channel', 'freq', 'segment')
```
Here is the corresponding repr `<Spectrum (from Raw) | 364 channels × 129 freqs × 651 segments, 0.0-300.3 Hz>`

~~Eventually (in another PR) I'd like to refactor the multitaper functions to make it possible to return the unaggregated tapers~~ **EDIT** *I think this was done already while I was on leave*, in which case you'd expect something like `raw.to_spectrum(multitaper, aggregate=False)._dims` to be `('channel', 'freq', 'taper')`.

`epochs.compute_psd()` and `evoked.compute_psd()` also work:

```py
events = mne.find_events(raw, stim_channel='STI 014')
event_dict = {'auditory/left': 1, 'auditory/right': 2, 'visual/left': 3}
epochs = mne.Epochs(raw, events, event_id=event_dict, preload=False,
                    proj=False)
evk = epochs.average()

epo_spec = epochs.compute_psd()
evk_spec = evk.compute_psd()
```

drammock commented:
this is still a draft but now would be a good time for some early feedback about code organization and API (ping @larsoner and @agramfort in particular, but input is welcome from anyone).

drammock commented:
> I am wondering about the use of spectrum. We use psd everywhere else.

I chose "spectrum" because it's more general: e.g., depending on what parameters people use, sometimes we plot the amplitude *spectrum*, or we plot 10*log10 of the power *spectrum*.

drammock commented:
Other reasons to pick a new name:

- people are used to things like `psds, freqs = psd_welch(raw)` so maybe (?) `raw.compute_psd()` might be similar enough to generate wrong expectations about the return value?
- having a more distinct name makes it less surprising that there are API changes.  For example the previous psd plotting functions have `area_mode` parameter which I'm proposing to change to `ci` in `spectrum.plot()` (`ci` is used in seaborn, and also in our own `plot_compare_evokeds`).

agramfort commented:
ok makes sense

> Message ID: ***@***.***>
>

dengemann commented:
@dengemann tagging myself to follow. Great this is alive!

dengemann commented:
Really happy to see this @drammock. Once basics are settled it would be cool to also add `plot_topomap`, based on `Epochs.plot_psd_topomap`. Not to forget adjusted `.to_data_frame`. Not suggesting all this should happen in this PR though :)

drammock commented on mne/time_frequency/tests/test_spectrum.py:
```diff
@@ -0,0 +1,188 @@
+    raw.compute_psd(**kwargs)
+
+
+@requires_h5py
+@pytest.mark.parametrize('inst', ('raw', 'epochs', 'evoked'))
+def test_spectrum_io(inst, tmp_path, request):
+    """Test save/load of spectrum objects."""
+    inst = request.getfixturevalue(inst)
+    fname = tmp_path / 'spectrum.h5'
+    orig = inst.compute_psd()
+    orig.save(fname)
+    loaded = read_spectrum(fname)
+    assert orig == loaded
```
@larsoner this test fails for evoked because of a limitation in pytest. The failure is:

```
_____________________________________________ test_spectrum_io[evoked] ______________________________________________
The requested fixture has no parameter defined for test:
    mne/time_frequency/tests/test_spectrum.py::test_spectrum_io[evoked]

Requested fixture '_evoked' defined in:
mne/conftest.py:306

Requested here:
/opt/miniforge3/envs/mnedev/lib/python3.10/site-packages/_pytest/fixtures.py:659
```

I can make it go away by removing `params=[testing._pytest_param()]` from the signature here:

https://github.com/mne-tools/mne-python/blob/2ba483ab91828be18a480e689d95f61c422ed6eb/mne/conftest.py#L305-L312

But presumably that will break your magic auto-skipping behavior.  The limitation is that `request.getfixturevalue()` cannot handle fixtures that have params, and this is [apparently by design](https://github.com/pytest-dev/pytest/issues/4666#issuecomment-456593913). A workaround is to use the [pytest-lazy-fixture](https://pypi.org/project/pytest-lazy-fixture) package.  Another alternative is to copy-paste a separate test for Evoked IO, like this:

```py
@requires_h5py
def test_evoked_spectrum_io(evoked, tmp_path):
    """Test save/load of spectrum objects."""
    fname = tmp_path / 'spectrum.h5'
    orig = evoked.compute_psd()
    orig.save(fname)
    loaded = read_spectrum(fname)
    assert orig == loaded
```
... which works because now the `evoked` fixture is listed directly in the test signature (instead of being accessed via `getfixturevalue()`), but is not very DRY.

What do you recommend?

larsoner replied:
I would work around it with:
```
@requires_h5py
@pytest.mark.parametrize('inst', ('raw', 'epochs', 'evoked'))
def test_spectrum_io(inst, tmp_path, request, evoked):
    """Test save/load of spectrum objects."""
    # work around parametrized fixtures not being accessible
    # via request.getfixturevalue
    # https://github.com/pytest-dev/pytest/issues/4666#issuecomment-456593913  # noqa: E501
    if inst == 'evoked':
        inst = evoked
    else:
        inst = request.getfixturevalue(inst)
    fname = tmp_path / 'spectrum.h5'
    orig = inst.compute_psd()
    orig.save(fname)
    loaded = read_spectrum(fname)
    assert orig == loaded
```

drammock replied:
Ah, OK. Inelegant but practical. Will do.

larsoner commented:
@drammock the bullet point is still unchecked:

>  support passing a callable to aggregate welch/multitaper estimates

Should we wait for a follow-up PR, or did you want to do it here?

Either way I can look at the code early tomorrow

drammock commented:
That was added to the list this week after conversations with @ajquinn. I think a follow up PR is probably best but if you disagree I can add it here. I don't think it will be too hard.

-------- Original Message --------
On Aug 25, 2022, 00:13, Eric Larson wrote:

> ***@***.***(https://github.com/drammock) the bullet point is still unchecked:
>
>> support passing a callable to aggregate welch/multitaper estimates
>
> Should we wait for a follow-up PR, or did you want to do it here?
>
> Either way I can look at the code early tomorrow
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/mne-tools/mne-python/pull/10184#issuecomment-1226586092), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AAN2AU3IGFK2KNBYFH2UMZDV22UDPANCNFSM5LNNZCNA).
> You are receiving this because you were mentioned.Message ID: ***@***.***>

larsoner commented on mne/viz/epochs.py:
```diff
@@ -982,19 +975,20 @@ def plot_epochs_psd(epochs, fmin=0, fmax=np.inf, tmin=None, tmax=None,
+        ax=ax, color=color, xscale=xscale, area_mode=area_mode,
+        area_alpha=area_alpha, dB=dB, estimate=estimate, show=show,
+        line_alpha=line_alpha, spatial_colors=spatial_colors, sphere=sphere,
+        exclude=exclude, n_jobs=n_jobs, average=average, verbose=verbose,
+        # these are **method_kw:
+        window='hamming', bandwidth=bandwidth, adaptive=adaptive,
+        low_bias=low_bias, normalization=normalization)
+    plt_show(show, fig)
```
Why not just pass `show=show` above, then you don't need `plt_show(show, fig)` here?

drammock replied:
good catch. actually already passing `show=show` above so this can just be removed.

larsoner commented on mne/viz/raw.py:
```diff
@@ -421,57 +410,47 @@ def plot_raw_psd(raw, fmin=0, fmax=np.inf, tmin=None, tmax=None, proj=False,
+        area_alpha=area_alpha, dB=dB, estimate=estimate, show=show,
+        line_alpha=line_alpha, spatial_colors=spatial_colors, sphere=sphere,
+        exclude=exclude, n_jobs=n_jobs, average=average, verbose=verbose,
+        n_fft=n_fft, n_overlap=n_overlap, window=window)
+    plt_show(show, fig)
```
Same here and maybe elsewhere

larsoner commented on mne/viz/utils.py:
```diff
@@ -2238,15 +2238,17 @@ def _plot_psd(inst, fig, freqs, psd_list, picks_list, titles_list,
-            else:  # area_mode is None
+            elif area_mode is None:
                 hyp_limits = None
+            else:  # area_mode is float
+                hyp_limits = _ci(psd)
```
Shouldn't you need to pass the `area_mode` value to `_ci` ? 
```suggestion
            else:  # area_mode is float
                hyp_limits = _ci(psd, area_mode)
```

larsoner commented on tutorials/epochs/20_visualize_epochs.py:
```diff
@@ -130,29 +130,47 @@
+#    The :meth:`~mne.Epochs.plot_psd` and `~mne.Epochs.plot_psd_topomap`
+#    methods of :class:`~mne.Epochs` objects are still provided to support
+#    legacy analysis scripts, but new code should instead use the
+#    :class:`~mne.time_frequency.EpochsSpectrum` object API.
```
It is weird to me to say this, then immediately use such legacy code below. Let's just switch to the new API below, too (and elsewhere). I think for legacy stuff where we keep around 2-line wrappers in order to avoid breaking people's code we can do this sort of thing.

drammock replied:
haha, I didn't mean to leave the old API in the tutorial. thanks for catching that.

larsoner commented on tutorials/raw/40_visualize_raw.py:
```diff
@@ -114,12 +114,25 @@
+#
+#    The :meth:`~mne.io.Raw.plot_psd` method of :class:`~mne.io.Raw` objects
+#    is still provided to support legacy analysis scripts, but new code should
+#    instead use the :class:`~mne.time_frequency.Spectrum` object API.
```
This one by contrast seems fine because the old API is *not* used immediately afterward

larsoner commented on mne/time_frequency/spectrum.py:
```diff
@@ -0,0 +1,1186 @@
+        from ..html_templates import repr_templates_env
+
+        inst_type_str = self._get_instance_type_string()
+        units = ', '.join(
+            f'{ch_type}: {unit}' for ch_type, unit in self.units().items())
```
Have you tried `'<br>'.join()` or something similar? I think it would be more readable and less risk of becoming really wide for a single row's data in 

https://output.circle-artifacts.com/output/job/b0cf67fb-c753-4c24-9ec2-0e9877d45f9e/artifacts/0/dev/auto_tutorials/time-freq/10_spectrum_class.html

drammock replied:
I had the same thought but `'<br/>'.join` did not work (the tag was rendered in the output).  I'll do a bit of searching to see if I can figure out another way to add linebreaks

larsoner commented on tutorials/time-freq/10_spectrum_class.py:
```diff
@@ -0,0 +1,172 @@
+# introduction of the :class:`~mne.time_frequency.Spectrum` and
+# :class:`~mne.time_frequency.EpochsSpectrum` classes.
+#
+# .. table:: Quick reference for common Spectral class actions
```
The table title is off-center for me

https://output.circle-artifacts.com/output/job/b0cf67fb-c753-4c24-9ec2-0e9877d45f9e/artifacts/0/dev/auto_tutorials/time-freq/10_spectrum_class.html#id4

<img width="810" alt="Screen Shot 2022-08-25 at 9 29 24 AM" src="https://user-images.githubusercontent.com/2365790/186615619-f8dc4d7a-62ed-487e-aa4b-fe4fc8750da7.png">


Can you either convert this to an issue, see if the fix is easy, or open an upstream pydata-sphinx-theme issue?

drammock replied:
it scrolls.

larsoner commented on mne/time_frequency/spectrum.py:
```diff
@@ -0,0 +1,1186 @@
+        psd_funcs = dict(welch=psd_array_welch,
+                         multitaper=psd_array_multitaper)
+        invalid_ix = np.in1d(list(method_kw),
+                             list(signature(psd_funcs[method]).parameters),
```
We typically use `_get_args` from `fixes` for this. Can you use it here and/or open an issue that we should investigate whether we can migrate to `inspect.signature` directly nowadays? I can't remember if we just needed to wait until 3.7+ or if `.parameters` isn't enough sometimes, or something else

drammock replied:
#11089

larsoner commented on mne/time_frequency/spectrum.py:
```diff
@@ -0,0 +1,1186 @@
+        self._data_type = ('Fourier Coefficients' if 'taper' in self._dims
+                           else 'Power Spectrum')
+
+    def __deepcopy__(self, memodict):
+        """Make a deepcopy."""
+        cls = self.__class__
+        result = cls.__new__(cls)
+        for k, v in self.__dict__.items():
+            v = deepcopy(v, memodict)
+            result.__dict__[k] = v
+        return result
```
Can you make a comment how this differs from the default implementation / why it's needed? To me it looks like the default and I'm a bit surprised you need it here at all

drammock replied:
originally I thought I had to override the one inherited from Epochs, but whatever the problem was it seems to be gone now.

larsoner commented on mne/time_frequency/spectrum.py:
```diff
@@ -0,0 +1,1186 @@
+        del self._inst_type
+        del other._inst_type
+        try:
+            same = object_diff(vars(self), vars(other)) == ''
+        finally:
+            self._inst_type = _inst_type
+            other._inst_type = _inst_type
```
Maybe a better approach would be to make `object_diff` handle classes? I think in principle something like this could work:
```
if inspect.isclass(a):
    if not inspect.isclass(b):
        same = False
    else:
        same = (a == b)
```

larsoner reviewed (commented):
Very clean! Just some minor comments from me, otherwise LGTM +1 for merge

wmvanvliet commented on doc/references.bib:
```diff
@@ -2285,6 +2285,19 @@ @article{LuckGaspelin2017
 
+@article{Welch1967,
+  title = {The Use of Fast {{Fourier}} Transform for the Estimation of Power Spectra: {{A}} Method Based on Time Averaging over Short, Modified Periodograms},
+  shorttitle = {The Use of Fast {{Fourier}} Transform for the Estimation of Power Spectra},
```
```suggestion
```

wmvanvliet commented on mne/viz/raw.py:
```diff
@@ -374,20 +371,12 @@ def plot_raw_psd(raw, fmin=0, fmax=np.inf, tmin=None, tmax=None, proj=False,
-        Number of points to use in Welch FFT calculations.
-        Default is None, which uses the minimum of 2048 and the
-        number of time points.
+        Number of points to use in Welch FFT calculations. Default is None,
```
```suggestion
        Number of points to use in Welch FFT calculations. Default is ``None``,
```

wmvanvliet commented on mne/viz/topomap.py:
```diff
@@ -1940,27 +1941,28 @@ def _plot_topomap_multi_cbar(data, pos, ax, title=None, unit=None, vmin=None,
+                            low_bias=True, normalization='length',
+                            ch_type=None, normalize=False, agg_fun=None,
+                            dB=False,  # sensors=True, show_names=False,
+                            # mask=None, mask_params=None, contours=6,
```
Why are these commented-out?

drammock replied:
these will be uncommented in a subsequent PR, where ALL the plot_topomap methods (for Cov, Proj, etc) will get their APIs standardized.

wmvanvliet commented on mne/viz/topomap.py:
```diff
@@ -1971,60 +1973,59 @@ def plot_epochs_psd_topomap(epochs, bands=None,
+    # add after vlim
+    # %(colorbar_topomap)s
+    # add after cbar_fmt
+    # %(units_topomap)s
```
This looks like some TODO but you want to hold off on it until topomap plotting has improved?

drammock commented:
- [new tutorial](https://output.circle-artifacts.com/output/job/e98a0c2d-c9ea-4c7c-adc5-f67ee86748aa/artifacts/0/dev/auto_tutorials/time-freq/10_spectrum_class.html#sphx-glr-auto-tutorials-time-freq-10-spectrum-class-py)
- changed tutorial [sleep stage classification](https://output.circle-artifacts.com/output/job/e98a0c2d-c9ea-4c7c-adc5-f67ee86748aa/artifacts/0/dev/auto_tutorials/clinical/60_sleep.html#sphx-glr-auto-tutorials-clinical-60-sleep-py)
- changed tutorial [raw plotting methods](https://output.circle-artifacts.com/output/job/e98a0c2d-c9ea-4c7c-adc5-f67ee86748aa/artifacts/0/dev/auto_tutorials/raw/40_visualize_raw.html#sphx-glr-auto-tutorials-raw-40-visualize-raw-py)
- changed tutorial [visualizing epochs](https://output.circle-artifacts.com/output/job/e98a0c2d-c9ea-4c7c-adc5-f67ee86748aa/artifacts/0/dev/auto_tutorials/epochs/20_visualize_epochs.html#sphx-glr-auto-tutorials-epochs-20-visualize-epochs-py)

wmvanvliet commented on tutorials/clinical/60_sleep.py:
```diff
@@ -188,19 +187,18 @@
 
 # iterate over the subjects
-stages = sorted(event_id.keys())
+stages = sorted(event_id)
```
```suggestion
stages = sorted(event_id.keys())
```
Explicit is better than implicit.

wmvanvliet commented on mne/time_frequency/spectrum.py:
```diff
@@ -0,0 +1,1166 @@
+from ..channels.channels import UpdateChannelsMixin, _get_ch_type
+from ..channels.layout import _merge_ch_data
+# from ..defaults import (_BORDER_DEFAULT, _EXTRAPOLATE_DEFAULT,
+#                         _INTERPOLATION_DEFAULT)
```
These commented-out things need to be addressed somehow sometime.

drammock replied:
these will be uncommented in a subsequent PR, where ALL the plot_topomap methods (for Cov, Proj, etc) will get their APIs standardized.

wmvanvliet reviewed (commented):
It's a work of art! 🤩 

Do we at some point want to support users creating their own `Spectral` instances like `EpochsArray` and `EvokedArray`?

drammock commented:
> Do we at some point want to support users creating their own `Spectral` instances like `EpochsArray` and `EvokedArray`?

this is something I struggled with, and ultimately decided "YAGNI".  But, I suspect it wouldn't be too difficult given that it's already basically implemented in the `_from_file` method for loading from disk.

drammock commented on mne/decoding/transformer.py:
```diff
@@ -12,7 +12,7 @@
 from .. import pick_types
 from ..filter import filter_data
-from ..time_frequency.psd import psd_array_multitaper
+from ..time_frequency import psd_array_multitaper
```
I don't understand why but it was at some point causing for me a circular import error.

drammock commented on mne/preprocessing/ica.py:
```diff
@@ -1648,8 +1647,8 @@ def find_bads_muscle(self, inst, threshold=0.5, start=None,
-        psd_func = psd_welch if isinstance(inst, BaseRaw) else psd_multitaper
-        psds, freqs = psd_func(sources, fmin=l_freq, fmax=h_freq, picks='misc')
+        spectrum = sources.compute_psd(fmin=l_freq, fmax=h_freq, picks='misc')
+        psds, freqs = spectrum.get_data(return_freqs=True)
```
yes.

wmvanvliet commented:
> ultimately decided "YAGNI"

Fair enough.

drammock commented on mne/time_frequency/tests/test_psd.py:
```diff
@@ -41,228 +34,126 @@ def test_psd_nan():
-            assert (np.abs(ixmax - ixtrue) < 2)
-
-        # Make sure the projection doesn't change channels it shouldn't
-        assert_array_almost_equal(psds, psds_proj)
```
yes. These old tests operated on the `psd_welch` and `psd_multitaper` funcs (which are now deprecated); now these tests only use `psd_array_[welch|multitaper]` (which don't take `proj`).  BUT testing with `proj` either True or False is still done, it's just done in `tests/test_spectrum.py` now.

drammock commented on mne/utils/docs.py:
```diff
@@ -1545,8 +1597,6 @@ def _reflow_param_docstring(docstring, has_first_line=True, width=75):
 """
 
-docdict['index_df'] = _index_df_base
-
```
nope. (if it was, CIs would have complained. Also I checked: `git grep "%(index_df)s"`)

larsoner reviewed (approved):
LGTM +1 for merge after the `_ci` fix!

larsoner commented:
https://output.circle-artifacts.com/output/job/79b51cb7-962f-4095-87ce-5aa72b49dcc1/artifacts/0/dev/auto_tutorials/time-freq/10_spectrum_class.html#sphx-glr-auto-tutorials-time-freq-10-spectrum-class-py

larsoner commented:
Congrats @drammock !

wmvanvliet commented:
Nice one!

agramfort commented:
🎉 🍻 👏 @drammock
